botocore/paginate.py

```diff
@@ -204,7 +208,16 @@ def _handle_first_request(self, parsed, primary_result_key,
         for token in self.result_keys:
             if token == primary_result_key:
                 continue
-            set_value_from_jmespath(parsed, token.expression, [])
+            sample = token.search(parsed)
+            if isinstance(sample, list):
+                empty_value = []
+            elif isinstance(sample, string_types):
+                empty_value = ''
+            elif isinstance(sample, (int, float)):
+                empty_value = 0
```

Comment:
While I don't have any particular objection for adding those 2 test cases (and I've updated this PR already), I would like to take this chance to say, I am worrying those new test cases and (especially) our tendency of "never break existing test cases", will further limit our future options on a potential paginator refactoring.

To better understand the background, let's imagine that we were still in day 1 of designing the paginator, it would be a matter of taste to decide whether an absent (primary or secondary) result_key should show up as the corresponding empty type (i.e. my previous [commit](https://github.com/rayluo/botocore/commit/ccdbcfc778ded1060e3db68f6f9e9c529afa0e14)), or not show up at all (i.e. my first [commit](https://github.com/rayluo/botocore/commit/a3e70de5e76f9f77806c959bb04ffe0ee8a04a3f)). In other words, these behavior was an implementation detail. And, implementation detail "should not impact API" in the first place. (There is a good explanation in slide 15 of this keynote http://lcsd05.cs.tamu.edu/slides/keynote.pdf and BTW, this is the webcast http://www.infoq.com/presentations/effective-api-design). If now it becomes a rigid part of our api contract and then limits our flexibility of refactoring, hence unnoticeably increases the complexity (thus the workload) on subsequent bug fix/new feature development, that doesn't feel right.

I'm NOT advocating it is OK to break any existing test cases. I'm just hoping we either: (a) do not put implementation detail into test cases, or; (b) or allow us to adjust those implementation level test cases when needed. Besides, we can make it clear that, all those response examples in documentation are merely demonstrating the format of each element, it does NOT hint each element will show up.

PS: This can be considered as a separated conversation. It should not delay or block the review of this PR.